File: C#/CSharpKatas/LuckyNumbers/LuckyNumbers.cs
using System;

namespace CSharpKatas.LuckyNumbers
{
    public class LuckyNumbers
    {
        public string GetLuckyNumbers(int range)
        {
            return PrintResult(FindLuckyNumbers(range));
        }

        public string GetLuckyPrimeNumbers(int range)
        {
            bool[] numbers = FindLuckyNumbers(range);

            for (int i = 0; i < numbers.Length; i++)
            {
                if (numbers[i]) continue;
                numbers[i] = !IsPrimeNumber(i+1);
            }

            return PrintResult(numbers);
        }

        private bool[] FindLuckyNumbers(int range)
        {
            if (range < 1) range = 0;
            bool[] numbers = new bool[range];
            int luckyCounter = 2;

            while (luckyCounter < numbers.Length)
            {
                luckyCounter = StrikeOutNumbers(numbers, luckyCounter);
            }

            return numbers;
        }

        private int StrikeOutNumbers(bool[] numbers, int luckyCounter)
        {
            int strikeCounter = 0;

            for (int i = 0; i < numbers.Length; i++)
            {
                if (numbers[i]) continue;
                strikeCounter++;
                
                if (strikeCounter == luckyCounter)
                {
                    numbers[i] = true;
                    strikeCounter = 0;
                }
            }

            return GetLuckyCounter(numbers, luckyCounter);
        }

        private int GetLuckyCounter(bool[] numbers, int skip)
        {
            if (skip >= numbers.Length) return numbers.Length;

            for (int i = skip; i < numbers.Length; i++)
            {
                if (!numbers[i]) return i+1;
            }

            return numbers.Length;
        }

        private static bool IsPrimeNumber(int number)
        {
            if (number == 1) return false;

            for (short i = 3; i <= Math.Sqrt(number); i += 2)
            {
                if (number % i == 0) return false;
            }

            return true;
        }

        private static string PrintResult(bool[] numbers)
        {
            string result = string.Empty;

            for (int i = 0; i < numbers.Length; i++)
            {
                if (!numbers[i]) result = result + (i+1) + ",";
            }

            if (!string.IsNullOrEmpty(result)) result = result.Substring(0, result.Length - 1);

            return result;
        }
    }
}

File: TypeScript/Katas/src/luckynumbers/luckynumbers.ts
export class LuckyNumbers {

    getLuckyNumbers(range: number): string {

        return this.toString(this.findLuckyNumbers(range));
    }

    getLuckyPrimeNumbers(range: number): string {

        var numbers = this.findLuckyNumbers(range);

        for (let i = 0; i < numbers.length; i++) {
            if (numbers[i]) continue;
            numbers[i] = !this.isPrime(i + 1);
        }

        return this.toString(numbers);
    }

    private findLuckyNumbers(range: number): boolean[] {

        var numbers = this.getNumbersArray(range);
        var luckyCounter = 2;

        while (luckyCounter < range) {
            luckyCounter = this.strikeOutNumbers(numbers, luckyCounter);
        }

        return numbers;
    }

    private strikeOutNumbers(numbers: boolean[], luckyCounter: number): number {

        var strikeCounter = 0;

        for (let i = 0; i < numbers.length; i++) {
            if (numbers[i]) continue;
            strikeCounter++;

            if (strikeCounter === luckyCounter) {
                numbers[i] = true;
                strikeCounter = 0;
            }
        }

        return this.getLuckyCounter(numbers, luckyCounter);
    }

    private getLuckyCounter(numbers: boolean[], luckyCounter: number): number {

        var numbersLength = numbers.length;
        if (luckyCounter >= numbersLength) return numbersLength;

        for (let i = luckyCounter; i < numbersLength; i++) {
            if (!numbers[i]) return i + 1;
        }

        return numbersLength;
    }

    private isPrime(num: number): boolean {

        if (num === 1) return false;

        for (let i = 3; i <= Math.sqrt(num); i += 2) {
            if (num % i === 0) return false;
        }

        return true;
    }

    private toString(numbers: boolean[]): string {

        var result = "";

        for (let i = 0; i < numbers.length; i++) {
            if (!numbers[i]) result += (i + 1) + ",";
        }

        if (result) result = result.substring(0, result.length - 1);

        return result;
    }

    private getNumbersArray(range: number): boolean[] {

        var numbers: boolean[] = [];
        for (let i = 0; i < range; i++) {
            numbers.push(false);
        }

        return numbers;
    }
}
